File: trunk/product/src/control/DC0080.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0080 Controller: Reception document line
 */

require_once(PATH_CTRL_FRMWK."/Controller.php");

class DC0080 extends Controller {


private function preQuery(&$params, &$where) {
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
    if (!empty($_REQUEST["QRY_MVMNTINDOC_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.MVMNTINDOC_ID like :MVMNTINDOC_ID";
      $params["MVMNTINDOC_ID"] = $_REQUEST["QRY_MVMNTINDOC_ID"];
    }
    if (!empty($_REQUEST["QRY_PRODUCT_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.PRODUCT_ID like :PRODUCT_ID";
      $params["PRODUCT_ID"] = $_REQUEST["QRY_PRODUCT_ID"];
    }
    if (!empty($_REQUEST["QRY_STOCKLOC_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "t.STOCKLOC_ID like :STOCKLOC_ID";
      $params["STOCKLOC_ID"] = $_REQUEST["QRY_STOCKLOC_ID"];
    }
}

public function doQuery() {
  try {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                t.ID
                ,t.LINE_NO
                ,t.MVMNTINDOC_ID
                ,t.PRODUCT_ID
                ,t.QTY
                ,t.BASE_DOC_QTY
                ,t.RECEIVED_QTY
                ,t.INVENTORY_QTY
                ,t.BASE_DOC_PRICE
                ,t.BASE_DOC_CURRENCY
                ,t.NOTES
                ,t.UOM
                ,pbo_product.get_name_by_id(t.product_id) PRODUCT_NAME
                ,t.STOCKLOC_ID
                ,pbo_org.get_stockloc_code_by_id(t.stockloc_id) STOCKLOC_CODE
            from MM_MOVEMENT_IN_LINE t $where $orderByClause ";
    $this->logger->debug($sql);
    $rs = $this->db->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->db->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $columns = array(
      "ID"
      ,"LINE_NO"
      ,"MVMNTINDOC_ID"
      ,"PRODUCT_ID"
      ,"QTY"
      ,"BASE_DOC_QTY"
      ,"RECEIVED_QTY"
      ,"INVENTORY_QTY"
      ,"BASE_DOC_PRICE"
      ,"BASE_DOC_CURRENCY"
      ,"NOTES"
      ,"UOM"
      ,"PRODUCT_NAME"
      ,"STOCKLOC_ID"
      ,"STOCKLOC_CODE"
      );
    $dataOut = $this->serializeCursor($rs,$columns, $this->query_data_format);
    if ($this->query_data_format == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doQuery  */


public function doExport() {
  try {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                t.ID
                ,t.LINE_NO
                ,t.MVMNTINDOC_ID
                ,t.PRODUCT_ID
                ,pbo_product.get_name_by_id(t.product_id) PRODUCT_NAME
                ,t.QTY
                ,t.UOM
                ,t.BASE_DOC_QTY
                ,t.RECEIVED_QTY
                ,t.INVENTORY_QTY
                ,t.BASE_DOC_PRICE
                ,t.BASE_DOC_CURRENCY
                ,pbo_org.get_stockloc_code_by_id(t.stockloc_id) STOCKLOC_CODE
                ,t.STOCKLOC_ID
                ,t.NOTES
            from MM_MOVEMENT_IN_LINE t $where $orderByClause ";
    $rs = $this->db->Execute($sql, $params);
    $rsCount = $this->db->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $columns = array(
     "ID"
     ,"LINE_NO"
     ,"MVMNTINDOC_ID"
     ,"PRODUCT_ID"
     ,"PRODUCT_NAME"
     ,"QTY"
     ,"UOM"
     ,"BASE_DOC_QTY"
     ,"RECEIVED_QTY"
     ,"INVENTORY_QTY"
     ,"BASE_DOC_PRICE"
     ,"BASE_DOC_CURRENCY"
     ,"STOCKLOC_CODE"
     ,"STOCKLOC_ID"
     ,"NOTES"
      );
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    }
    $dataOut = $this->serializeCursor($rs,$columns,"xml");
    $dataOut = "<records>".$dataOut."</records>";
    $dataOut = "<queryParams>".$this->serializeArray($params,"xml")."</queryParams>".$dataOut;
    $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fieldDef,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fieldDef,false)."</columnDef>".$dataOut;
    $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
    $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
    $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    $this->beginExport();
    print $dataOut;
    $this->endExport();
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doExport  */


public function fetchRecord() {
  try {
    $RECORD = array();
    $RECORD["ID"] = $this->getRequestParam("ID");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0080.fetchRecord().");}
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function fetchRecord */


public function doInsert() {
  $this->logger->debug("start: ".$this->dcName.".doInsert");
  try {
    $RECORD = array();
    $RECORD["BASE_DOC_CURRENCY"] = $this->getRequestParam("BASE_DOC_CURRENCY");
    $RECORD["BASE_DOC_PRICE"] = $this->getRequestParam("BASE_DOC_PRICE");
    $RECORD["BASE_DOC_QTY"] = $this->getRequestParam("BASE_DOC_QTY");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["INVENTORY_QTY"] = $this->getRequestParam("INVENTORY_QTY");
    $RECORD["LINE_NO"] = $this->getRequestParam("LINE_NO");
    $RECORD["MVMNTINDOC_ID"] = $this->getRequestParam("MVMNTINDOC_ID");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["PRODUCT_ID"] = $this->getRequestParam("PRODUCT_ID");
    $RECORD["PRODUCT_NAME"] = $this->getRequestParam("PRODUCT_NAME");
    $RECORD["QTY"] = $this->getRequestParam("QTY");
    $RECORD["RECEIVED_QTY"] = $this->getRequestParam("RECEIVED_QTY");
    $RECORD["STOCKLOC_CODE"] = $this->getRequestParam("STOCKLOC_CODE");
    $RECORD["STOCKLOC_ID"] = $this->getRequestParam("STOCKLOC_ID");
    $RECORD["UOM"] = $this->getRequestParam("UOM");
    $sql = "insert into MM_MOVEMENT_IN_LINE(
                 ID
                ,LINE_NO
                ,MVMNTINDOC_ID
                ,PRODUCT_ID
                ,QTY
                ,BASE_DOC_QTY
                ,RECEIVED_QTY
                ,INVENTORY_QTY
                ,BASE_DOC_PRICE
                ,BASE_DOC_CURRENCY
                ,NOTES
                ,UOM
                ,STOCKLOC_ID
            ) values ( 
                 :ID
                ,:LINE_NO
                ,:MVMNTINDOC_ID
                ,:PRODUCT_ID
                ,:QTY
                ,:BASE_DOC_QTY
                ,:RECEIVED_QTY
                ,:INVENTORY_QTY
                ,:BASE_DOC_PRICE
                ,:BASE_DOC_CURRENCY
                ,:NOTES
                ,:UOM
                ,:STOCKLOC_ID
    )";
    $stmt = $this->db->prepare($sql);
    $_seq = $this->db->execute("select SEQ_MVMNTINLIN_ID.nextval seq_val from dual")->fetchRow();
    $RECORD["ID"] = $_seq["SEQ_VAL"];
    $this->logger->debug("insert of RECORD: ".$this->logger->map2string($RECORD) );
    $this->db->Execute($stmt, $RECORD);
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("end: ".$this->dcName.".doInsert");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doInsert */


public function doUpdate() {
  $this->logger->debug("Start: ".$this->dcName.".doUpdate");
  try {
    $RECORD = array();
    $RECORD["BASE_DOC_CURRENCY"] = $this->getRequestParam("BASE_DOC_CURRENCY");
    $RECORD["BASE_DOC_PRICE"] = $this->getRequestParam("BASE_DOC_PRICE");
    $RECORD["BASE_DOC_QTY"] = $this->getRequestParam("BASE_DOC_QTY");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["INVENTORY_QTY"] = $this->getRequestParam("INVENTORY_QTY");
    $RECORD["LINE_NO"] = $this->getRequestParam("LINE_NO");
    $RECORD["MVMNTINDOC_ID"] = $this->getRequestParam("MVMNTINDOC_ID");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["PRODUCT_ID"] = $this->getRequestParam("PRODUCT_ID");
    $RECORD["PRODUCT_NAME"] = $this->getRequestParam("PRODUCT_NAME");
    $RECORD["QTY"] = $this->getRequestParam("QTY");
    $RECORD["RECEIVED_QTY"] = $this->getRequestParam("RECEIVED_QTY");
    $RECORD["STOCKLOC_CODE"] = $this->getRequestParam("STOCKLOC_CODE");
    $RECORD["STOCKLOC_ID"] = $this->getRequestParam("STOCKLOC_ID");
    $RECORD["UOM"] = $this->getRequestParam("UOM");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0080.doUpdate().");}
    $sql = "update MM_MOVEMENT_IN_LINE set 
                 ID=:ID
                ,LINE_NO=:LINE_NO
                ,MVMNTINDOC_ID=:MVMNTINDOC_ID
                ,PRODUCT_ID=:PRODUCT_ID
                ,QTY=:QTY
                ,BASE_DOC_QTY=:BASE_DOC_QTY
                ,RECEIVED_QTY=:RECEIVED_QTY
                ,INVENTORY_QTY=:INVENTORY_QTY
                ,BASE_DOC_PRICE=:BASE_DOC_PRICE
                ,BASE_DOC_CURRENCY=:BASE_DOC_CURRENCY
                ,NOTES=:NOTES
                ,UOM=:UOM
                ,STOCKLOC_ID=:STOCKLOC_ID
    where 
           ID= :ID
    ";
    $stmt = $this->db->prepare($sql);
    $this->logger->debug("update of RECORD: ".$this->logger->map2string($RECORD) );
    $this->db->Execute($stmt, $RECORD);
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".doUpdate");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doUpdate */


public function doDelete() {
  $this->logger->debug("Start: ".$this->dcName.".doDelete");
  try {
    $RECORD["ID"] = $this->getRequestParam("ID");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0080.doDelete().");}
    $sql = "delete from MM_MOVEMENT_IN_LINE where 
           ID= :ID
    ";
    $stmt = $this->db->prepare($sql);
    $this->db->Execute($stmt, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".doDelete");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doDelete */


public function initNewRecord() {
  try {
    $RECORD["BASE_DOC_CURRENCY"] = $this->getRequestParam("BASE_DOC_CURRENCY");
    $RECORD["BASE_DOC_PRICE"] = $this->getRequestParam("BASE_DOC_PRICE");
    $RECORD["BASE_DOC_QTY"] = $this->getRequestParam("BASE_DOC_QTY");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["INVENTORY_QTY"] = $this->getRequestParam("INVENTORY_QTY");
    $RECORD["LINE_NO"] = $this->getRequestParam("LINE_NO");
    $RECORD["MVMNTINDOC_ID"] = $this->getRequestParam("MVMNTINDOC_ID");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["PRODUCT_ID"] = $this->getRequestParam("PRODUCT_ID");
    $RECORD["PRODUCT_NAME"] = $this->getRequestParam("PRODUCT_NAME");
    $RECORD["QTY"] = $this->getRequestParam("QTY");
    $RECORD["RECEIVED_QTY"] = $this->getRequestParam("RECEIVED_QTY");
    $RECORD["STOCKLOC_CODE"] = $this->getRequestParam("STOCKLOC_CODE");
    $RECORD["STOCKLOC_ID"] = $this->getRequestParam("STOCKLOC_ID");
    $RECORD["UOM"] = $this->getRequestParam("UOM");
    $RECORD["_p_record_status"] = "insert";
    $this->setFieldInitialValues($RECORD, "DC0080");
    print "{success:true, data:".json_encode($RECORD)."}";
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                t.ID
                ,t.LINE_NO
                ,t.MVMNTINDOC_ID
                ,t.PRODUCT_ID
                ,t.QTY
                ,t.BASE_DOC_QTY
                ,t.RECEIVED_QTY
                ,t.INVENTORY_QTY
                ,t.BASE_DOC_PRICE
                ,t.BASE_DOC_CURRENCY
                ,t.NOTES
                ,t.UOM
                ,pbo_product.get_name_by_id(t.product_id) PRODUCT_NAME
                ,t.STOCKLOC_ID
                ,pbo_org.get_stockloc_code_by_id(t.stockloc_id) STOCKLOC_CODE
            from MM_MOVEMENT_IN_LINE t
         where 
           t.ID= :ID
            ";
    $rs = $this->db->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    $this->copyArrayValues($record, $row);
} /* end function findByPk  */

private  $fieldDef = array(
  "ID" => array("DATA_TYPE" => "NUMBER")
  ,"LINE_NO" => array("DATA_TYPE" => "NUMBER")
  ,"MVMNTINDOC_ID" => array("DATA_TYPE" => "NUMBER")
  ,"PRODUCT_ID" => array("DATA_TYPE" => "NUMBER")
  ,"QTY" => array("DATA_TYPE" => "NUMBER")
  ,"BASE_DOC_QTY" => array("DATA_TYPE" => "NUMBER")
  ,"RECEIVED_QTY" => array("DATA_TYPE" => "NUMBER")
  ,"INVENTORY_QTY" => array("DATA_TYPE" => "NUMBER")
  ,"BASE_DOC_PRICE" => array("DATA_TYPE" => "NUMBER")
  ,"BASE_DOC_CURRENCY" => array("DATA_TYPE" => "STRING")
  ,"NOTES" => array("DATA_TYPE" => "STRING")
  ,"UOM" => array("DATA_TYPE" => "STRING")
  ,"PRODUCT_NAME" => array("DATA_TYPE" => "STRING")
  ,"STOCKLOC_ID" => array("DATA_TYPE" => "NUMBER")
  ,"STOCKLOC_CODE" => array("DATA_TYPE" => "STRING")
);


private function readRequest(&$RECORD) {
     if (isset($_REQUEST["BASE_DOC_CURRENCY"] )) { $RECORD["BASE_DOC_CURRENCY"] = $this->getRequestParam("BASE_DOC_CURRENCY"); }
     if (isset($_REQUEST["BASE_DOC_PRICE"] )) { $RECORD["BASE_DOC_PRICE"] = $this->getRequestParam("BASE_DOC_PRICE"); }
     if (isset($_REQUEST["BASE_DOC_QTY"] )) { $RECORD["BASE_DOC_QTY"] = $this->getRequestParam("BASE_DOC_QTY"); }
     if (isset($_REQUEST["ID"] )) { $RECORD["ID"] = $this->getRequestParam("ID"); }
     if (isset($_REQUEST["INVENTORY_QTY"] )) { $RECORD["INVENTORY_QTY"] = $this->getRequestParam("INVENTORY_QTY"); }
     if (isset($_REQUEST["LINE_NO"] )) { $RECORD["LINE_NO"] = $this->getRequestParam("LINE_NO"); }
     if (isset($_REQUEST["MVMNTINDOC_ID"] )) { $RECORD["MVMNTINDOC_ID"] = $this->getRequestParam("MVMNTINDOC_ID"); }
     if (isset($_REQUEST["NOTES"] )) { $RECORD["NOTES"] = $this->getRequestParam("NOTES"); }
     if (isset($_REQUEST["PRODUCT_ID"] )) { $RECORD["PRODUCT_ID"] = $this->getRequestParam("PRODUCT_ID"); }
     if (isset($_REQUEST["PRODUCT_NAME"] )) { $RECORD["PRODUCT_NAME"] = $this->getRequestParam("PRODUCT_NAME"); }
     if (isset($_REQUEST["QTY"] )) { $RECORD["QTY"] = $this->getRequestParam("QTY"); }
     if (isset($_REQUEST["RECEIVED_QTY"] )) { $RECORD["RECEIVED_QTY"] = $this->getRequestParam("RECEIVED_QTY"); }
     if (isset($_REQUEST["STOCKLOC_CODE"] )) { $RECORD["STOCKLOC_CODE"] = $this->getRequestParam("STOCKLOC_CODE"); }
     if (isset($_REQUEST["STOCKLOC_ID"] )) { $RECORD["STOCKLOC_ID"] = $this->getRequestParam("STOCKLOC_ID"); }
     if (isset($_REQUEST["UOM"] )) { $RECORD["UOM"] = $this->getRequestParam("UOM"); }
} /* end function readRequest  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
  try {
    $RECORD = array();
    $this->readRequest($RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
  }catch(Exception  $e) {
      System::sendActionErrorJson( $e->getMessage());
  }
} /* end function callProcedure */

}
?>

File: ServerJava/src/net/nan21/ebs/dc/DC0080.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0080 DC Controller: Reception document line
 */

package net.nan21.ebs.dc;


import java.sql.Connection;
import java.sql.ResultSet;
import java.util.Date;
import java.util.HashMap;
import java.util.Properties;
import javax.servlet.http.HttpServletResponse;
import net.nan21.ebs.lib.CollectionUtils;
import net.nan21.ebs.lib.AbstractDataControl;
import net.nan21.ebs.lib.FieldDef;
import net.nan21.ebs.lib.HttpRequest;
import net.nan21.ebs.lib.HttpSession;
import net.nan21.ebs.lib.IDataControl;
import net.nan21.ebs.lib.DbManager;

public class DC0080 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_MVMNTINDOC_ID") != null && !this.request.getParam("QRY_MVMNTINDOC_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.MVMNTINDOC_ID like :MVMNTINDOC_ID");
      this.queryParams.put("MVMNTINDOC_ID",(String)this.request.getParam("QRY_MVMNTINDOC_ID"));
    }
    if (this.request.getParam("QRY_PRODUCT_ID") != null && !this.request.getParam("QRY_PRODUCT_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.PRODUCT_ID like :PRODUCT_ID");
      this.queryParams.put("PRODUCT_ID",(String)this.request.getParam("QRY_PRODUCT_ID"));
    }
    if (this.request.getParam("QRY_STOCKLOC_ID") != null && !this.request.getParam("QRY_STOCKLOC_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("t.STOCKLOC_ID like :STOCKLOC_ID");
      this.queryParams.put("STOCKLOC_ID",(String)this.request.getParam("QRY_STOCKLOC_ID"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " t.BASE_DOC_CURRENCY"+
               " ,t.BASE_DOC_PRICE"+
               " ,t.BASE_DOC_QTY"+
               " ,t.ID"+
               " ,t.INVENTORY_QTY"+
               " ,t.LINE_NO"+
               " ,t.MVMNTINDOC_ID"+
               " ,t.NOTES"+
               " ,t.PRODUCT_ID"+
               " ,pbo_product.get_name_by_id(t.product_id) PRODUCT_NAME"+
               " ,t.QTY"+
               " ,t.RECEIVED_QTY"+
               " ,pbo_org.get_stockloc_code_by_id(t.stockloc_id) STOCKLOC_CODE"+
               " ,t.STOCKLOC_ID"+
               " ,t.UOM"+
           " from MM_MOVEMENT_IN_LINE t "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " t.ID"+
               " ,t.LINE_NO"+
               " ,t.MVMNTINDOC_ID"+
               " ,t.PRODUCT_ID"+
               ",pbo_product.get_name_by_id(t.product_id) PRODUCT_NAME"+
               " ,t.QTY"+
               " ,t.UOM"+
               " ,t.BASE_DOC_QTY"+
               " ,t.RECEIVED_QTY"+
               " ,t.INVENTORY_QTY"+
               " ,t.BASE_DOC_PRICE"+
               " ,t.BASE_DOC_CURRENCY"+
               ",pbo_org.get_stockloc_code_by_id(t.stockloc_id) STOCKLOC_CODE"+
               " ,t.STOCKLOC_ID"+
               " ,t.NOTES"+
           " from MM_MOVEMENT_IN_LINE t "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into MM_MOVEMENT_IN_LINE("+
               "  BASE_DOC_CURRENCY"+
               " ,BASE_DOC_PRICE"+
               " ,BASE_DOC_QTY"+
               " ,ID"+
               " ,INVENTORY_QTY"+
               " ,LINE_NO"+
               " ,MVMNTINDOC_ID"+
               " ,NOTES"+
               " ,PRODUCT_ID"+
               " ,QTY"+
               " ,RECEIVED_QTY"+
               " ,STOCKLOC_ID"+
               " ,UOM"+
           " ) values ( "+
               "  :BASE_DOC_CURRENCY"+
               " ,:BASE_DOC_PRICE"+
               " ,:BASE_DOC_QTY"+
               " ,:ID"+
               " ,:INVENTORY_QTY"+
               " ,:LINE_NO"+
               " ,:MVMNTINDOC_ID"+
               " ,:NOTES"+
               " ,:PRODUCT_ID"+
               " ,:QTY"+
               " ,:RECEIVED_QTY"+
               " ,:STOCKLOC_ID"+
               " ,:UOM"+
    ")";
    this.record.put("ID",   dbm.getSequenceNextValue("SEQ_MVMNTINLIN_ID")  );
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update MM_MOVEMENT_IN_LINE set "+
               "  ID=:ID"+
               " ,LINE_NO=:LINE_NO"+
               " ,MVMNTINDOC_ID=:MVMNTINDOC_ID"+
               " ,PRODUCT_ID=:PRODUCT_ID"+
               " ,QTY=:QTY"+
               " ,BASE_DOC_QTY=:BASE_DOC_QTY"+
               " ,RECEIVED_QTY=:RECEIVED_QTY"+
               " ,INVENTORY_QTY=:INVENTORY_QTY"+
               " ,BASE_DOC_PRICE=:BASE_DOC_PRICE"+
               " ,BASE_DOC_CURRENCY=:BASE_DOC_CURRENCY"+
               " ,NOTES=:NOTES"+
               " ,UOM=:UOM"+
               " ,STOCKLOC_ID=:STOCKLOC_ID"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from MM_MOVEMENT_IN_LINE where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
               " t.BASE_DOC_CURRENCY"+
               " ,t.BASE_DOC_PRICE"+
               " ,t.BASE_DOC_QTY"+
               " ,t.ID"+
               " ,t.INVENTORY_QTY"+
               " ,t.LINE_NO"+
               " ,t.MVMNTINDOC_ID"+
               " ,t.NOTES"+
               " ,t.PRODUCT_ID"+
                ",pbo_product.get_name_by_id(t.product_id) PRODUCT_NAME"+
               " ,t.QTY"+
               " ,t.RECEIVED_QTY"+
                ",pbo_org.get_stockloc_code_by_id(t.stockloc_id) STOCKLOC_CODE"+
               " ,t.STOCKLOC_ID"+
               " ,t.UOM"+
           " from MM_MOVEMENT_IN_LINE t"+
        " where "+
     "      t.ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void callProcedure(String pName)  throws Exception {
    this.populateRecordFromRequest();
}


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("BASE_DOC_CURRENCY", new FieldDef("STRING"));
	  this.fields.put("BASE_DOC_PRICE", new FieldDef("NUMBER"));
	  this.fields.put("BASE_DOC_QTY", new FieldDef("NUMBER"));
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("INVENTORY_QTY", new FieldDef("NUMBER"));
	  this.fields.put("LINE_NO", new FieldDef("NUMBER"));
	  this.fields.put("MVMNTINDOC_ID", new FieldDef("NUMBER"));
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  this.fields.put("PRODUCT_ID", new FieldDef("NUMBER"));
	  this.fields.put("PRODUCT_NAME", new FieldDef("STRING"));
	  this.fields.put("QTY", new FieldDef("NUMBER"));
	  this.fields.put("RECEIVED_QTY", new FieldDef("NUMBER"));
	  this.fields.put("STOCKLOC_CODE", new FieldDef("STRING"));
	  this.fields.put("STOCKLOC_ID", new FieldDef("NUMBER"));
	  this.fields.put("UOM", new FieldDef("STRING"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	}

public void doCustomAction(String action) {}
}
